library/core/src/num/nonzero.rs

```diff
@@ -287,6 +288,550 @@ nonzero_integers_div! {
+                #[doc = concat!("let max = ", stringify!($Ty), "::new(",
+                                stringify!($Int), "::MAX)?;")]
+                ///
+                /// assert_eq!(Some(four), two.checked_mul(two));
+                /// assert_eq!(None, max.checked_mul(two));
+                /// # Ok(())
+                /// # }
+                /// ```
+                #[unstable(feature = "nonzero_ops", issue = "84186")]
+                #[inline]
+                pub const fn checked_mul(self, other: $Ty) -> Option<$Ty> {
+                    if let Some(result) = self.get().checked_mul(other.get()) {
+                        // SAFETY: checked_mul returns None on overflow
+                        // and `other` is also non-null
+                        // so the result cannot be zero.
+                        Some(unsafe { $Ty::new_unchecked(result) })
+                    } else {
+                        None
+                    }
+                }
+
+                /// Multiply two non-zero integers together.
+                #[doc = concat!("Return [`", stringify!($Int), "::MAX`] on overflow.")]
+                ///
+                /// # Examples
+                ///
+                /// ```
+                /// #![feature(nonzero_ops)]
+                /// # #![feature(try_trait)]
+                #[doc = concat!("# use std::num::", stringify!($Ty), ";")]
+                ///
+                /// # fn main() -> Result<(), std::option::NoneError> {
+                #[doc = concat!("let two = ", stringify!($Ty), "::new(2)?;")]
+                #[doc = concat!("let four = ", stringify!($Ty), "::new(4)?;")]
+                #[doc = concat!("let max = ", stringify!($Ty), "::new(",
+                                stringify!($Int), "::MAX)?;")]
+                ///
+                /// assert_eq!(four, two.saturating_mul(two));
+                /// assert_eq!(max, four.saturating_mul(max));
+                /// # Ok(())
+                /// # }
+                /// ```
+                #[unstable(feature = "nonzero_ops", issue = "84186")]
+                #[inline]
+                pub const fn saturating_mul(self, other: $Ty) -> $Ty {
+                    // SAFETY: saturating_mul returns u*::MAX on overflow
+                    // and `other` is also non-null
+                    // so the result cannot be zero.
+                    unsafe { $Ty::new_unchecked(self.get().saturating_mul(other.get())) }
+                }
+
+                /// Multiply two non-zero integers together,
+                /// assuming overflow cannot occur.
+                /// Overflow is unchecked, and it is undefined behaviour to overflow
+                /// *even if the result would wrap to a non-zero value*.
+                /// The behaviour is undefined as soon as
+                #[doc = concat!("`self * rhs > ", stringify!($Int), "::MAX`, ")]
+                #[doc = concat!("or `self * rhs < ", stringify!($Int), "::MIN`.")]
```

Comment:
This is only possible in the signed case, but the macro generates the same doc for both signed and unsigned types. Should I attempt to generate a different signed/unsigned doc for this specific line (splitting the macro, some kind of additional/conditional information like `$(signed)?`?), or is it ok to leave as-is?

Reply:
Oh, macros. Fun.

You could do this:

Create a macro `signed_or_unsigned($ident, signed: expr, unsigned: expr)` that expands to one or the other expression depending on the first ident:

```rust
macro_rules! signed_or_unsigned {
    (signed, signed: $signed:expr, unsigned: $unsigned:expr) => {
        $signed
    };
    (unsigned, signed: $signed:expr, unsigned: $unsigned:expr) => {
        $unsigned
    };
}
```

Then add a `unsigned` or `signed` word in the `nonzero_unsigned_signed_operations` for every type:

```rust
nonzero_unsigned_signed_operations! {
    unsigned NonZeroU8(u8);
    unsigned NonZeroU16(u16);
    unsigned NonZeroU32(u32);
    unsigned NonZeroU64(u64);
    unsigned NonZeroU128(u128);
    unsigned NonZeroUsize(usize);
    signed NonZeroI8(i8);
    signed NonZeroI16(i16);
    signed NonZeroI32(i32);
    signed NonZeroI64(i64);
    signed NonZeroI128(i128);
    signed NonZeroIsize(isize);
}
```

Capture that as `$signed:ident`:

```rust
macro_rules! nonzero_unsigned_signed_operations {
    ( $( $signed:ident $Ty: ident($Int: ty); )+ ) => {
```

And then use `#[doc = signed_or_unsigned!($signed, ..)]` like this:

```rust
                #[doc = signed_or_unsigned!(
                    $signed,
                    signed: concat!(...),
                    unsigned: concat!(...)
                )]
```

Reply:
Oh I see, you're emulating some kind of dedicated ternary operator.. I'll use your hint. Thank you :)